Parsing improvements

aelij opened the pull request:
* Using a tokenizer for parsing XAML strings (e.g. `Point` and `Matrix`) - uses far less allocations and faster 
* Faster lookup & cache of known colors/brushes names

aelij committed:
Using a tokenizer instead of string split

aelij committed:
Improve known color/brush parsing

aelij committed:
Merge branch 'master' into tokenizer

grokys committed:
Merge branch 'master' into tokenizer

Commit:
Merge branch 'master' into tokenizer

aelij committed:
Merge branch 'master' into tokenizer

grokys committed:
Merge branch 'master' into tokenizer

grokys commented on src/Avalonia.Controls/Avalonia.Controls.csproj:
```diff
@@ -26,6 +26,7 @@
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
   </PropertyGroup>
   <ItemGroup>
+    <Compile Include="..\Avalonia.Visuals\Utilities\StringTokenizer.cs" Link="Utils\StringTokenizer.cs" />
```
I think I'd prefer to make `StringTokenizer` public rather than include it separately in different projects, it seems rather hackish like this? It should probably be moved into `Avalonia.Base\Utilities`.

grokys commented on src/Avalonia.Visuals/Media/Brush.cs:
```diff
@@ -34,26 +34,21 @@ public double Opacity
         /// <returns>The <see cref="Color"/>.</returns>
         public static IBrush Parse(string s)
         {
+            if (s == null) throw new ArgumentNullException(nameof(s));
```
Elsewhere we use `Contract.Requires` for argument null checks.

grokys commented on src/Avalonia.Visuals/Utilities/StringTokenizer.cs:
```diff
@@ -0,0 +1,205 @@
+            return success;
+        }
+
+        public int NextInt32Required(char? separator = null)
```
Would it make more sense to name these methods using the standard reader naming, e.g. `ReadInt32`, `TryReadInt32`?

grokys reviewed (changes requested):
👍 in general, just a few tweaks I think, let me know if you disagree. There are also some failing tests.

grokys committed:
Merge branch 'master' into pr/1125-parsing-improvements

grokys committed:
Moved StringTokenizer to Avalonia.Base.

And make it public instead of linking it from Avalonia.Visuals to Avalonia.Controls.

grokys committed:
Use Contract.Requires for preconditions.

grokys committed:
Use standard naming for Read methods.

grokys committed:
Merge branch 'master' into tokenizer

grokys commented:
@aelij looks like you don't have time to finish this off, so I made the changes I suggested. Do they look ok to you?

grokys committed:
Fix failing tests.

aelij commented:
@grokys Sorry, I kinda forgot about this one :) Looks good
